sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/processor/CompactBlockProcessor.kt

```diff
@@ -183,31 +206,108 @@ class CompactBlockProcessor internal constructor(
+                maxDelayMillis = MAX_BACKOFF_INTERVAL
+            ) {
                 val result = processingMutex.withLockLogged("processNewBlocks") {
-                    processNewBlocks()
+                    when (subTreeRootResult) {
+                        is GetSubtreeRootsResult.SpendBeforeSync -> {
+                            // Pass the commitment tree data to the database
+                            when (
+                                val result = putSaplingSubtreeRoots(
+                                    backend = backend,
+                                    startIndex = 0,
+                                    subTreeRootList = (subTreeRootResult as GetSubtreeRootsResult.SpendBeforeSync)
+                                        .subTreeRootList,
+                                    lastValidHeight = lowerBoundHeight
+                                )
+                            ) {
+                                PutSaplingSubtreeRootsResult.Success -> {
+                                    // Lets continue with the next step
+                                }
+                                is PutSaplingSubtreeRootsResult.Failure -> {
+                                    BlockProcessingResult.SyncFailure(result.failedAtHeight, result.exception)
+                                }
+                            }
+                            processNewBlocksInSbSOrder(
+                                backend = backend,
+                                downloader = downloader,
+                                repository = repository,
+                                network = network,
+                                lastValidHeight = lowerBoundHeight,
+                                firstUnenhancedHeight = _processorInfo.value.firstUnenhancedHeight
+                            )
+                        }
+                        GetSubtreeRootsResult.Linear -> {
+                            // This is caused by an empty response result. Although the spend-before-sync
+                            // synchronization algorithm is not supported, we can get the entire block range as we
+                            // previously did for the linear sync type.
+                            processNewBlocksInSbSOrder(
+                                backend = backend,
+                                downloader = downloader,
+                                repository = repository,
+                                network = network,
+                                lastValidHeight = lowerBoundHeight,
+                                firstUnenhancedHeight = _processorInfo.value.firstUnenhancedHeight
+                            )
+                        }
+                        is GetSubtreeRootsResult.OtherFailure -> {
+                            // The server possibly replied with some unsupported error. We still approach
+                            // spend-before-sync synchronization.
```

Comment:
```suggestion
                        GetSubtreeRootsResult.Linear, is GetSubtreeRootsResult.OtherFailure -> {
                            // This is caused by an empty response result or another unsupported error.
                            // Although the spend-before-sync synchronization algorithm is not supported,
                            // we can get the entire block range as we previously did for the linear sync type.
```
(Assuming this syntax works; if it doesn't then never mind.) Non-blocking.

Reply:
Yep, it works like this. Thanks for pointing this out. I just used the shorter version and merged their comments.